experimental/dds/tree2/docs/schema2.md

```diff
@@ -0,0 +1,129 @@
+# Schema 2
+
+Proposal for changes to the [schema system](../src/core/schema-stored/Stored%20and%20View%20Schema.md).
+
+## Motivation
+
+The current schema system has several issues:
+
+1. It's cluttered with features that are not used in most cases. (ex: extraLocalFields, extraGlobalFields, global fields and values are not used on almost all schema)
+2. It's not clear to users which features when and why they all exist.
+3. global value fields and extra global fields don't interact in a clear well defined way (would such value fields be required on all nodes with extraGlobalFields or are they implicitly made optional when used via extra global fields?)
+4. Extra local fields can result in types TypeScript can't model, since our schema system only applies the schema for the extra fields to all fields not explicitly listed. This [can't be done in TypeScript](https://www.typescriptlang.org/play?noPropertyAccessFromIndexSignature=true&ts=4.5.5#code/PTAEBUCcE9QFwPagLYEMDWBTUr7QA7YDuAFppNgGYCWmANgCagBEqzo1AzqJ3JNQDsA5gBocAphV44KoAQFdkAI3IAuAFDqQoACIJM3AQjigiCSOnVwC2AMrVk+OtgC8oAN7qAkAG0ssQR4+QSEAXVVQf1BMAA84TAluVnYAfiD+YVAIhWVyAG51AF8CrTA9AzljU3N0DWtCUFtg4U4AeTgySHASVAEAHnBouISGbl4MoQA+UDd-BEoPUD9MAIFQAApBgB90kIBKcMiVofjEiFA0gUwAN3Iso9hiqxtG5qE2jvJu3oAmAZORmM3tM3ABRGIAYzo8gYmD64xCYnAkxK2gAQsYSPAyJxsKhZAAreTSAAGCOEJOeDVsM1eEw+nW+-WSKKpdh+tKa9PajJ6Aj+LNRZX0hiqSnk1EY6kE8UglFQELsDicmA5ni8qAi5KEBV8UUCnEOckUKkgBSepV0IsqJnFkoY0oEsvlisayucau8mt2wl1y1WjSNOVN5pKEIQAmkAH0GKg4KgAIwReyOZy09w4CLMGLsJ7hyMmGNx1A-ZPu1yLb3Z5hiJQRACsoCe2gAAnBOABaWKECFwLuQSDmUyDzL504meqYdT56Ox+MAZjLqYrGe9jebYDbne7mF7-cHkGHEaEoDHCQnNmnEdnxYALEuVenMywc7Ws9Bc7rNF4gA)
+5. It requires use of symbols to avoid colliding with local field names in many places in the API, which has been confusing users (many TypeScript developers don't even know what Symbols are).
+6. Supporting all the things in all the nodes makes APIs complex. This is extra annoying since this complicates the schema-aware API.
+7. Extra fields require proxies to implement the desired TypeScript APIs for nodes.
+8. Having both global and local fields adds complexity, particularly to schema and path storage where symbols can't be used.
+9. Supporting extraLocalFields and extraGlobalFields on nodes while providing a JavaScript object like API requires a Proxy, which adds overhead and complicates code implementation and maintenance.
+
+## Proposal
+
+Instead of one kind of node in view schema, have 4, each with a subset of our current functionality:
+
+1. Terminal Node: holds a value, but no children.
+   Don't support `undefined` as a value: an empty Struct can be used for that case if needed.
+2. Struct Node: finite list of fields (key+field type).
+3. Map Node: node with a single single field type which must permit empty.
```

Comment:
I assume the "single single" in an unintended repetition.

I don't really understand what "single field type which must permit empty" means here. Are you saying that Map nodes would be parameterized with a field kind whose multiplicity would have to tolerate zero children? For example, one could make a map of optionals, or a map of sequences. If that is what you mean then maybe "field kind" would be clearer since "field type" could be referring to the type of children allowed in a field.

Reply:
should be more clear/fixed now.